adding grpc.start to logging interceptors

domgreen opened the pull request:
This pull request will add a `grpc.start_time` to the context logger showing when the gRPC call started. This used the [RFC3339](https://tools.ietf.org/html/rfc3339) format to store the data as a `string` in the context.

mwitkow commented:
a) Next time, please write a comment describing what `grpc.start` is. First comment on PR is super useful for context of interview.

b) Please rename `grpc.start` to `grpc.start_time` and specify, and test it's format. I know we're implictly using whatever time formatting the logging library uses, but it would be great if we either standardized it or documented it.

c) Please document the `grpc.start_time` or actually, even better, all known fields that are emitted in JSON and their format.

artsiukhou commented on logging/logrus/payload_interceptors_test.go:
```diff
@@ -73,41 +72,33 @@ func (s *logrusPayloadSuite) getServerAndClientMessages(expectedServer int, expe
 
 func (s *logrusPayloadSuite) TestPingError_LogsOnlyRequestsOnError() {
 	_, err := s.Client.PingError(s.SimpleCtx(), &pb_testproto.PingRequest{Value: "something", ErrorCodeReturned: uint32(4)})
 	require.Error(s.T(), err, "there must be not be an on a successful call")
```
Wrong error message.

artsiukhou commented on logging/logrus/payload_interceptors_test.go:
```diff
@@ -128,9 +119,11 @@ func (s *logrusPayloadSuite) TestPingStream_LogsAllRequestsAndResponses() {
-		assert.Contains(s.T(), m, `.content": {`, "all messages must contain payloads")
+		assert.Equal(s.T(), m["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+		assert.Equal(s.T(), m["grpc.method"], "PingStream", "all lines must contain method name")
+		assert.Equal(s.T(), m["level"], "info", "all lines must contain method name")
```
Wrong error message.

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -50,19 +50,22 @@ func (s *logrusServerSuite) TestPing_WithCustomTags() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
```
I think we should change check here and 2 lines above (`NoError`) to `require` otherway it can cause runtime errors in the code below.

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -166,16 +177,18 @@ func (s *logrusServerOverrideSuite) TestPing_HasOverriddenDuration() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
```
`require`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -192,16 +205,18 @@ func (s *logrusServerOverrideSuite) TestPingList_HasOverriddenDuration() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
```
`require`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -192,16 +205,18 @@ func (s *logrusServerOverrideSuite) TestPingList_HasOverriddenDuration() {
+	assert.Equal(s.T(), msgs[1]["msg"], "finished streaming call", "handler's message must contain user message")
+	assert.Equal(s.T(), msgs[1]["level"], "info", "OK error codes must be logged on info level.")
+	assert.NotContains(s.T(), msgs[1], "grpc.time_ms", "handler's message must not contain default duration")
+	assert.Contains(s.T(), msgs[1], "grpc.duration", "handler's message must not contain overridden duration")
```
Rm `not`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -240,9 +255,9 @@ func (s *logrusServerOverrideDeciderSuite) TestPing_HasOverriddenDecider() {
 	msgs := s.getOutputJSONs()
 	assert.Len(s.T(), msgs, 1, "single log statements should be logged")
```
`require`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -258,10 +273,10 @@ func (s *logrusServerOverrideDeciderSuite) TestPingError_HasOverriddenDecider()
-	assert.Contains(s.T(), m, fmt.Sprintf(`"level": "%s"`, level.String()), msg)
+	assert.Equal(s.T(), m["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+	assert.Equal(s.T(), m["grpc.method"], "PingError", "all lines must contain method name")
+	assert.Equal(s.T(), m["grpc.code"], code.String(), "all lines must contain method name")
```
Wrong error message.

artsiukhou commented on logging/zap/client_interceptors_test.go:
```diff
@@ -76,13 +72,13 @@ func (s *zapClientSuite) TestPingList() {
+	assert.Equal(s.T(), msgs[0]["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+	assert.Equal(s.T(), msgs[0]["grpc.method"], "PingList", "all lines must contain method name")
+	assert.Equal(s.T(), msgs[0]["msg"], "finished client streaming call", "handler's message must contain user message")
+	assert.Contains(s.T(), msgs[0], "grpc.time_ms", "handler's message must not contain default duration")
```
Wrong error message, rm `not`

artsiukhou commented on logging/zap/client_interceptors_test.go:
```diff
@@ -119,11 +115,11 @@ func (s *zapClientSuite) TestPingError_WithCustomLevels() {
+		assert.Equal(s.T(), msgs[0]["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+		assert.Equal(s.T(), msgs[0]["grpc.method"], "PingError", "all lines must contain method name")
+
+		assert.Equal(s.T(), msgs[0]["grpc.code"], tcase.code.String(), "all lines must contain method name")
```
Wrong error message.

artsiukhou commented on logging/zap/client_interceptors_test.go:
```diff
@@ -152,14 +148,11 @@ func (s *zapClientOverrideSuite) TestPing_HasOverrides() {
+	assert.Equal(s.T(), msgs[0]["grpc.method"], "Ping", "all lines must contain method name")
+	assert.Equal(s.T(), msgs[0]["msg"], "finished client unary call", "handler's message must contain user message")
+	assert.NotContains(s.T(), msgs[0], "grpc.time_ms", "handler's message must not contain default duration")
+	assert.Contains(s.T(), msgs[0], "grpc.duration", "handler's message must not contain overridden duration")
```
Wrong error message, rm `not`.

artsiukhou commented on logging/zap/client_interceptors_test.go:
```diff
@@ -174,12 +167,11 @@ func (s *zapClientOverrideSuite) TestPingList_HasOverrides() {
+	assert.Equal(s.T(), msgs[0]["grpc.method"], "PingList", "all lines must contain method name")
+	assert.Equal(s.T(), msgs[0]["msg"], "finished client streaming call", "handler's message must contain user message")
+	assert.NotContains(s.T(), msgs[0], "grpc.time_ms", "handler's message must not contain default duration")
+	assert.Contains(s.T(), msgs[0], "grpc.duration", "handler's message must not contain overridden duration")
```
`must contain`

artsiukhou commented on logging/zap/payload_interceptors_test.go:
```diff
@@ -71,42 +70,34 @@ func (s *zapPayloadSuite) getServerAndClientMessages(expectedServer int, expecte
 func (s *zapPayloadSuite) TestPing_LogsBothRequestAndResponse() {
-	resp, err := s.Client.Ping(s.SimpleCtx(), goodPing)
+	_, err := s.Client.Ping(s.SimpleCtx(), goodPing)
 	assert.NoError(s.T(), err, "there must be not be an on a successful call")
```
`require`

artsiukhou commented on logging/zap/payload_interceptors_test.go:
```diff
@@ -71,42 +70,34 @@ func (s *zapPayloadSuite) getServerAndClientMessages(expectedServer int, expecte
-	assert.Contains(s.T(), serverResp, fmt.Sprintf(`"counter": %d`, resp.Counter))
+	assert.Contains(s.T(), clientReq, "grpc.request.content", "request payload must be logged in a structured way")
+	assert.Contains(s.T(), serverReq, "grpc.request.content", "request payload must be logged in a structured way")
+	assert.Contains(s.T(), clientResp, "grpc.response.content", "request payload must be logged in a structured way")
```
Fix error message (`request` -> `response`).

artsiukhou commented on logging/zap/payload_interceptors_test.go:
```diff
@@ -71,42 +70,34 @@ func (s *zapPayloadSuite) getServerAndClientMessages(expectedServer int, expecte
+	assert.Contains(s.T(), clientReq, "grpc.request.content", "request payload must be logged in a structured way")
+	assert.Contains(s.T(), serverReq, "grpc.request.content", "request payload must be logged in a structured way")
+	assert.Contains(s.T(), clientResp, "grpc.response.content", "request payload must be logged in a structured way")
+	assert.Contains(s.T(), serverResp, "grpc.response.content", "request payload must be logged in a structured way")
```
Fix error message (`request` -> `response`)

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -61,18 +61,22 @@ func (s *zapServerSuite) TestPing_WithCustomTags() {
+		assert.NotNil(s.T(), m["custom_tags.int"], "all lines must contain `custom_tags.int` set by AddFields")
+		assert.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+		_, err := time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -110,10 +114,15 @@ func (s *zapServerSuite) TestPingError_WithCustomLevels() {
+		assert.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+
+		_, err = time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -131,17 +140,21 @@ func (s *zapServerSuite) TestPingList_WithCustomTags() {
+		assert.Equal(s.T(), m["grpc.request.value"], "something", "all lines must contain fields extracted from goodPing because of test.manual_extractfields.pb")
+		require.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+		_, err := time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -50,19 +50,22 @@ func (s *logrusServerSuite) TestPing_WithCustomTags() {
+		assert.NotNil(s.T(), m["custom_tags.int"], "all lines must contain `custom_tags.int` set by AddFields")
+		assert.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+		_, err := time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -100,10 +103,15 @@ func (s *logrusServerSuite) TestPingError_WithCustomLevels() {
+		assert.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+
+		_, err = time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/logrus/server_interceptors_test.go:
```diff
@@ -121,18 +129,21 @@ func (s *logrusServerSuite) TestPingList_WithCustomTags() {
+		assert.Equal(s.T(), m["grpc.request.value"], "something", "all lines must contain fields extracted from goodPing because of test.manual_extractfields.pb")
+		require.NotNil(s.T(), m["grpc.start_time"], "all lines must contain the start time")
+		_, err := time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
```
`assert.NoError`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -61,18 +61,22 @@ func (s *zapServerSuite) TestPing_WithCustomTags() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
```
`require`

artsiukhou commented on logging/zap/payload_interceptors_test.go:
```diff
@@ -71,42 +70,34 @@ func (s *zapPayloadSuite) getServerAndClientMessages(expectedServer int, expecte
 
 func (s *zapPayloadSuite) TestPingError_LogsOnlyRequestsOnError() {
 	_, err := s.Client.PingError(s.SimpleCtx(), &pb_testproto.PingRequest{Value: "something", ErrorCodeReturned: uint32(4)})
 	require.Error(s.T(), err, "there must be not be an on a successful call")
```
Wrong error message.

artsiukhou commented on logging/zap/payload_interceptors_test.go:
```diff
@@ -127,9 +118,11 @@ func (s *zapPayloadSuite) TestPingStream_LogsAllRequestsAndResponses() {
-		assert.Contains(s.T(), m, `.content": {`, "all messages must contain payloads")
+		assert.Equal(s.T(), m["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+		assert.Equal(s.T(), m["grpc.method"], "PingStream", "all lines must contain method name")
+		assert.Equal(s.T(), m["level"], "info", "all lines must contain method name")
```
Wrong error message.

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -110,10 +114,15 @@ func (s *zapServerSuite) TestPingError_WithCustomLevels() {
+		_, err = time.Parse(time.RFC3339, m["grpc.start_time"].(string))
+		assert.Nil(s.T(), err, "should be able to parse date as RFC3339")
+
+		assert.Equal(s.T(), m["grpc.code"], tcase.code.String(), "all lines must contain method name")
```
Wrong error message.

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -175,16 +188,17 @@ func (s *zapServerOverrideSuite) TestPing_HasOverriddenDuration() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
```
`require`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -175,16 +188,17 @@ func (s *zapServerOverrideSuite) TestPing_HasOverriddenDuration() {
+	assert.Equal(s.T(), msgs[1]["msg"], "finished unary call", "handler's message must contain user message")
+	assert.Equal(s.T(), msgs[1]["level"], "info", "OK error codes must be logged on info level.")
+	assert.NotContains(s.T(), msgs[1], "grpc.time_ms", "handler's message must not contain default duration")
+	assert.Contains(s.T(), msgs[1], "grpc.duration", "handler's message must not contain overridden duration")
```
`must contain`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -201,16 +215,18 @@ func (s *zapServerOverrideSuite) TestPingList_HasOverriddenDuration() {
 	assert.Len(s.T(), msgs, 2, "two log statements should be logged")
 	for _, m := range msgs {
 		s.T()
```
Seems like we don't need this here and in all other tests.

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -248,9 +264,9 @@ func (s *zapServerOverridenDeciderSuite) TestPing_HasOverriddenDecider() {
 	msgs := s.getOutputJSONs()
 	assert.Len(s.T(), msgs, 1, "single log statements should be logged")
```
`require`

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -266,10 +282,10 @@ func (s *zapServerOverridenDeciderSuite) TestPingError_HasOverriddenDecider() {
-	assert.Contains(s.T(), m, fmt.Sprintf(`"level": "%s"`, level.String()), msg)
+	assert.Equal(s.T(), m["grpc.service"], "mwitkow.testproto.TestService", "all lines must contain service name")
+	assert.Equal(s.T(), m["grpc.method"], "PingError", "all lines must contain method name")
+	assert.Equal(s.T(), m["grpc.code"], code.String(), "all lines must contain method name")
```
Must contain code.

artsiukhou commented on logging/zap/server_interceptors_test.go:
```diff
@@ -284,9 +300,9 @@ func (s *zapServerOverridenDeciderSuite) TestPingList_HasOverriddenDecider() {
 	}
 	msgs := s.getOutputJSONs()
 	assert.Len(s.T(), msgs, 1, "single log statements should be logged")
```
`require`.

domgreen commented:
@artsiukhou fixed up all the test messages and added `require` where appropriate and changed the check on parsing from `Nil` to `NoError`.

@mwitkow have changed the naming of the tag and added some more documentation.

mwitkow commented:
Legit!
